File: Drag&Drop/Demos/DoubleCollectionViewController.m
//
//  DoubleCollectionViewController.m
//  Drag&Drop
//
//  Created by Joey Patino on 11/14/15.
//  Copyright © 2015 Joseph Patino. All rights reserved.
//

#import "DoubleCollectionViewController.h"
#import "UICollectionView+DragDropControllerSupport.h"

@interface DoubleCollectionViewController () <UICollectionViewDataSource, UICollectionViewDelegate, UICollectionViewDelegateFlowLayout>

@property (nonatomic, strong) UICollectionView *leftCollectionView;
@property (nonatomic, strong) UICollectionViewFlowLayout *leftLayout;
@property (nonatomic, strong) NSMutableArray *leftDataSource;


@property (nonatomic, strong) UICollectionView *rightCollectionView;
@property (nonatomic, strong) UICollectionViewFlowLayout *rightLayout;
@property (nonatomic, strong) NSMutableArray *rightDataSource;

@end

@implementation DoubleCollectionViewController

- (void)viewDidLoad {
    [super viewDidLoad];
    [self loadContent];
}

- (void)loadContent {
    [self loadLeftContent];
    [self loadRightContent];

}

- (void)loadLeftContent {
    self.leftDataSource = [NSMutableArray array];
    for (int i = 0; i < 8; i++) {
        [self.leftDataSource addObject:@(i)];
    }

    self.leftLayout = [[UICollectionViewFlowLayout alloc] init];
    self.leftLayout.minimumInteritemSpacing = 4;
    self.leftLayout.minimumLineSpacing = 4;
    self.leftLayout.sectionInset = UIEdgeInsetsMake(4, 4, 4, 4);
    self.leftCollectionView = [[UICollectionView alloc] initWithFrame:CGRectMake(0, 00,
                                                                                 CGRectGetWidth(self.view.frame)/2,
                                                                                 CGRectGetHeight(self.view.frame)-60)
                               
                                                 collectionViewLayout:self.leftLayout];
    self.leftCollectionView.delegate = self;
    self.leftCollectionView.dataSource = self;
    [self.view addSubview:self.leftCollectionView];
    [self.leftCollectionView registerClass:[UICollectionViewCell class] forCellWithReuseIdentifier:@"CollectionViewCell"];
    
    [self.leftCollectionView reloadData];
}

- (void)loadRightContent {
    self.rightDataSource = [NSMutableArray array];
    for (int i = 4; i < 12; i++) {
        [self.rightDataSource addObject:@(i)];
    }

    self.rightLayout = [[UICollectionViewFlowLayout alloc] init];
    self.rightLayout.minimumInteritemSpacing = 4;
    self.rightLayout.minimumLineSpacing = 4;
    self.rightLayout.sectionInset = UIEdgeInsetsMake(4, 4, 4, 4);
    self.rightCollectionView = [[UICollectionView alloc] initWithFrame:CGRectMake(CGRectGetWidth(self.view.frame)/2, 00,
                                                                                 CGRectGetWidth(self.view.frame)/2,
                                                                                 CGRectGetHeight(self.view.frame)-60)
                               
                                                 collectionViewLayout:self.rightLayout];
    self.rightCollectionView.delegate = self;
    self.rightCollectionView.dataSource = self;
    [self.view addSubview:self.rightCollectionView];
    [self.rightCollectionView registerClass:[UICollectionViewCell class] forCellWithReuseIdentifier:@"CollectionViewCell"];
    
    [self.rightCollectionView reloadData];
}

#pragma mark -

- (NSInteger)collectionView:(UICollectionView *)collectionView numberOfItemsInSection:(NSInteger)section {

    if (collectionView == self.leftCollectionView) return self.leftDataSource.count;
    
    return self.rightDataSource.count;
}

- (NSInteger)numberOfSectionsInCollectionView:(UICollectionView *)collectionView {
    return 1;
}

#pragma mark -

- (UICollectionViewCell *)collectionView:(UICollectionView *)collectionView cellForItemAtIndexPath:(NSIndexPath *)indexPath {
    
    UICollectionViewCell *cell = [collectionView dequeueReusableCellWithReuseIdentifier:@"CollectionViewCell" forIndexPath:indexPath];
    [collectionView enableDragAndDropForCell:cell];

    cell.backgroundColor = [UIColor whiteColor];
    [cell.contentView.subviews makeObjectsPerformSelector:@selector(removeFromSuperview)];
    
    NSNumber *n = nil;
    if (collectionView == self.rightCollectionView)
        n = self.rightDataSource[indexPath.row];
    else
        n = self.leftDataSource[indexPath.row];
    
    [self applyLabel:[NSString stringWithFormat:@"%li", (long)n.integerValue] toView:cell.contentView];
    
    return cell;
}

#pragma mark -

- (void)collectionView:(UICollectionView *)collectionView moveItemAtIndexPath:(NSIndexPath *)sourceIndexPath
      toCollectionView:(UICollectionView *)destinationCollectionView toIndexPath:(NSIndexPath *)destinationIndexPath {
    DLog(@"%s", __PRETTY_FUNCTION__);
    
    id item = nil;
    if ([collectionView isEqual:self.leftCollectionView]) {
        item = [self.leftDataSource objectAtIndex:sourceIndexPath.row];
        [self.leftDataSource removeObject:item];
    }
    else if ([collectionView isEqual:self.rightCollectionView]) {
        item = [self.rightDataSource objectAtIndex:sourceIndexPath.row];
        [self.rightDataSource removeObject:item];
    }
    
    if ([destinationCollectionView isEqual:self.rightCollectionView]) {
        [self.rightDataSource insertObject:item atIndex:destinationIndexPath.row];
    }
    else if ([destinationCollectionView isEqual:self.leftCollectionView]) {
        [self.leftDataSource insertObject:item atIndex:destinationIndexPath.row];
    }
    
    NSLog(@"L: %@", self.leftDataSource);
    NSLog(@"R: %@", self.rightDataSource);
}

- (BOOL)collectionView:(UICollectionView *)collectionView canMoveItemAtIndexPath:(NSIndexPath *)indexPath
      toCollectionView:(UICollectionView *)destinationCollectionView toIndexPath:(NSIndexPath *)toIndexPath {
    DLog(@"%s", __PRETTY_FUNCTION__);

    NSArray *source = nil;
    NSNumber *obj = nil;
    
    if ([collectionView isEqual:self.leftCollectionView]){
        source = self.rightDataSource;
        obj = self.leftDataSource[indexPath.row];
    }
    else if ([collectionView isEqual:self.rightCollectionView]) {
        source = self.leftDataSource;
        obj = self.rightDataSource[indexPath.row];
    }

    NSUInteger idx = [source indexOfObjectPassingTest:^BOOL(NSNumber *number, NSUInteger idx, BOOL *stop){
        return (number.integerValue == obj.integerValue);
    }];

    if (idx == NSNotFound)
        return YES;

    return NO;
}

- (BOOL)collectionView:(UICollectionView *)collectionView canMoveItemAtIndexPath:(NSIndexPath *)indexPath {
    DLog(@"%s", __PRETTY_FUNCTION__);
//    if (indexPath.row == 0) return NO;
    
    return YES;
}

- (void)collectionView:(UICollectionView *)collectionView moveItemAtIndexPath:(NSIndexPath *)sourceIndexPath toIndexPath:(NSIndexPath*)destinationIndexPath {
    DLog(@"%s", __PRETTY_FUNCTION__);
    
    if (collectionView == self.leftCollectionView) {
        NSObject *item = [self.leftDataSource objectAtIndex:sourceIndexPath.row];
        [self.leftDataSource removeObject:item];
        [self.leftDataSource insertObject:item atIndex:destinationIndexPath.row];
    }
    else if (collectionView == self.rightCollectionView) {
        NSObject *item = [self.rightDataSource objectAtIndex:sourceIndexPath.row];
        [self.rightDataSource removeObject:item];
        [self.rightDataSource insertObject:item atIndex:destinationIndexPath.row];
    }
}

#pragma mark -

- (CGSize)collectionView:(UICollectionView *)collectionView layout:(UICollectionViewLayout *)collectionViewLayout sizeForItemAtIndexPath:(NSIndexPath *)indexPath {
    
    return CGSizeMake((collectionView.bounds.size.width/2)-6, 120);
}

#pragma mark -

- (void)applyLabel:(NSString *)string toView:(UIView *)view {
    UILabel *l = [[UILabel alloc] init];
    l.text = string;
    [l sizeToFit];
    l.frame = CGRectMake(0, 0, CGRectGetWidth(l.frame), CGRectGetHeight(l.frame));
    [view addSubview:l];
}

@end

File: Demo/DragDropDemo/Demos/DoubleCollectionViewController.swift
//
//  DoubleCollectionViewController.swift
//  DragDropDemo
//
//  Created by Joey Patino on 11/14/15.
//  Copyright © 2015 Joseph Patino. All rights reserved.
//

import UIKit
import DragDrop
import DemoKit

/// Two collection views, and a move between them that the datasource can veto:
/// a player already on the destination list cannot be moved there again.
final class DoubleCollectionViewController: DemoViewController {

    private var leftCollectionView: UICollectionView?
    private var leftDataSource: [Player] = []
    private var leftPanel: PanelView?

    private var rightCollectionView: UICollectionView?
    private var rightDataSource: [Player] = []
    private var rightPanel: PanelView?

    override func loadContent() {
        title = "Lineup"

        leftDataSource = Array(SampleData.players.prefix(8))
        rightDataSource = Array(SampleData.players.suffix(8))

        // Each side is its own titled card. Two bare collection views side by
        // side read as one grid with a line down it; a card with its own
        // header, hue and count reads as two lists, which is what they are.
        let (leftPanel, leftGrid) = installSide(title: "Starters",
                                                subtitle: "On the pitch",
                                                symbol: "figure.soccer",
                                                hue: .mint,
                                                identifier: "starters",
                                                x: 0)
        self.leftPanel = leftPanel
        self.leftCollectionView = leftGrid

        let (rightPanel, rightGrid) = installSide(title: "Bench",
                                                 subtitle: "Available",
                                                 symbol: "chair.lounge.fill",
                                                 hue: .slate,
                                                 identifier: "bench",
                                                 x: view.bounds.width / 2)
        self.rightPanel = rightPanel
        self.rightCollectionView = rightGrid

        leftGrid.reloadData()
        rightGrid.reloadData()
        refreshCounts()
    }

    private func installSide(title: String,
                             subtitle: String,
                             symbol: String,
                             hue: DemoTheme.Hue,
                             identifier: String,
                             x: CGFloat) -> (PanelView, UICollectionView) {

        let top = contentTop + 8
        let panel = PanelView(title: title, subtitle: subtitle, symbolName: symbol, hue: hue)
        let frame = CGRect(x: x + 8, y: top,
                           width: view.bounds.width / 2 - 16,
                           height: view.bounds.height - top - view.safeAreaInsets.bottom - 8)
        let content = install(panel, in: view, frame: frame)

        let layout = UICollectionViewFlowLayout()
        layout.minimumInteritemSpacing = 6
        layout.minimumLineSpacing = 6
        layout.sectionInset = UIEdgeInsets(top: 2, left: 0, bottom: 8, right: 0)

        let grid = UICollectionView(frame: content.bounds, collectionViewLayout: layout)
        grid.autoresizingMask = [.flexibleWidth, .flexibleHeight]
        grid.backgroundColor = .clear
        grid.delegate = self
        grid.dataSource = self
        grid.accessibilityIdentifier = identifier
        grid.register(UICollectionViewCell.self, forCellWithReuseIdentifier: "CollectionViewCell")
        content.addSubview(grid)

        // The grid is the drop target, not the card's content view that holds
        // it, so the owning panel has to be named rather than looked up. No
        // controller to pass: a collection view builds its own internally.
        register(grid, in: panel)

        // The grid fills the card, so the card is what should light up -- the
        // same accepting state the panel screens use, rather than a border
        // drawn around a scroll view inside it.
        grid.dropHighlight = { [weak panel] accepting in
            panel?.setDropState(accepting ? .accepting : .idle)
        }

        return (panel, grid)
    }

    private func refreshCounts() {
        leftPanel?.count = leftDataSource.count
        rightPanel?.count = rightDataSource.count
    }
}

// MARK: -

extension DoubleCollectionViewController: UICollectionViewDelegate, UICollectionViewDelegateFlowLayout {

    func collectionView(_ collectionView: UICollectionView,
                        layout collectionViewLayout: UICollectionViewLayout,
                        sizeForItemAt indexPath: IndexPath) -> CGSize {
        CGSize(width: (collectionView.bounds.size.width / 2) - 12, height: 120)
    }
}

// MARK: -

extension DoubleCollectionViewController: UICollectionViewDataSourceCellSwapSupport {

    func collectionView(_ collectionView: UICollectionView, numberOfItemsInSection section: Int) -> Int {
        if collectionView === leftCollectionView { return leftDataSource.count }
        return rightDataSource.count
    }

    func numberOfSections(in collectionView: UICollectionView) -> Int { 1 }

    // MARK: -

    func collectionView(_ collectionView: UICollectionView,
                        cellForItemAt indexPath: IndexPath) -> UICollectionViewCell {

        let cell = collectionView.dequeueReusableCell(withReuseIdentifier: "CollectionViewCell",
                                                      for: indexPath)
        collectionView.enableDragAndDrop(for: cell)

        cell.backgroundColor = .clear
        cell.contentView.subviews.forEach { $0.removeFromSuperview() }

        let player = collectionView === rightCollectionView
            ? rightDataSource[indexPath.row]
            : leftDataSource[indexPath.row]

        let card = PlayerCardView(player: player, identifier: "player-\(player.id)")
        card.frame = cell.contentView.bounds
        card.autoresizingMask = [.flexibleWidth, .flexibleHeight]
        cell.contentView.addSubview(card)

        return cell
    }

    // MARK: -

    func collectionView(_ collectionView: UICollectionView,
                        moveItemAt sourceIndexPath: IndexPath,
                        to destinationCollectionView: UICollectionView,
                        to destinationIndexPath: IndexPath) {

        var moved: Player?
        if collectionView === leftCollectionView {
            moved = leftDataSource.remove(at: sourceIndexPath.row)
        } else if collectionView === rightCollectionView {
            moved = rightDataSource.remove(at: sourceIndexPath.row)
        }

        guard let moved else { return }

        if destinationCollectionView === rightCollectionView {
            rightDataSource.insert(moved, at: min(destinationIndexPath.row, rightDataSource.count))
        } else if destinationCollectionView === leftCollectionView {
            leftDataSource.insert(moved, at: min(destinationIndexPath.row, leftDataSource.count))
        }

        refreshCounts()
    }

    /// A player already on the destination list cannot be moved there again.
    func collectionView(_ collectionView: UICollectionView,
                        canMoveItemAt indexPath: IndexPath,
                        to destinationCollectionView: UICollectionView,
                        to toIndexPath: IndexPath) -> Bool {

        let destination: [Player]
        let player: Player

        if collectionView === leftCollectionView {
            destination = rightDataSource
            player = leftDataSource[indexPath.row]
        } else if collectionView === rightCollectionView {
            destination = leftDataSource
            player = rightDataSource[indexPath.row]
        } else {
            return true
        }

        return !destination.contains(player)
    }

    func collectionView(_ collectionView: UICollectionView, canMoveItemAt indexPath: IndexPath) -> Bool {
        true
    }

    func collectionView(_ collectionView: UICollectionView,
                        moveItemAt sourceIndexPath: IndexPath,
                        to destinationIndexPath: IndexPath) {

        if collectionView === leftCollectionView {
            let player = leftDataSource.remove(at: sourceIndexPath.row)
            leftDataSource.insert(player, at: destinationIndexPath.row)
        } else if collectionView === rightCollectionView {
            let player = rightDataSource.remove(at: sourceIndexPath.row)
            rightDataSource.insert(player, at: destinationIndexPath.row)
        }
    }
}
